spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,800 @@
+                        .addClassAndMethod(this));
+            }
+            synchronizedMethods.add(xMethod);
+        }
+
+        SignatureParser signature = new SignatureParser(xMethod.getSignature());
+        if ("readResolve".equals(obj.getName()) && "()Ljava/lang/Object;".equals(obj.getSignature())
+                && Subtypes2.instanceOf(getThisClass(), "java.io.Serializable")) {
+            return;
+        }
+
+        if (!(xMethod.isStatic() || xMethod.isPublic())) {
+            return;
+        }
+
+        String returnType = signature.getReturnTypeSignature();
+        String javaStyleClassType = "L" + getClassName() + ";";
+        if (returnType.equals(javaStyleClassType)) {
+            if ("clone".equals(obj.getName())) {
+                return;
+            }
+            exposingMethods.add(xMethod);
+        } else {
+            String sourceSig = xMethod.getSourceSignature();
+            if (sourceSig != null) {
+                String typePattern = "^.*<.*" + javaStyleClassType + ".*>;$";
+                GenericSignatureParser genericSignature = new GenericSignatureParser(sourceSig);
+                String genericReturnValue = genericSignature.getReturnTypeSignature();
+                if (genericReturnValue.matches(typePattern)) {
+                    exposingMethods.add(xMethod);
+                }
+            }
+        }
+
+    }
+
+    @Override
+    public void visit(Field obj) {
+        XField xField = getXField();
+        if (isCollection(xField)) {
+            try {
+                if (isInherited(xField)) {
+                    if (!xField.isPrivate()) {
+                        badBackingCollections.add(xField);
+                    }
+
+                    findExposureInHierarchy(xField)
+                            .forEach(method -> collectionAccessorsInHierarchy.add(xField, method));
+                } else {
+                    if (xField.isPublic()) {
+                        badBackingCollections.add(xField);
+                    }
+
+                    if (xField.isProtected() && !getThisClass().isFinal() || isPackagePrivate(xField)) {
+                        inheritableBackingCollections.add(xField);
+                    }
+                }
+            } catch (ClassNotFoundException e) {
+                AnalysisContext.reportMissingClass(e);
+            }
+        }
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.MONITORENTER && stack.getStackDepth() > 0) {
+            XField lock = stack.getStackItem(0).getXField();
+
+            if (lock != null) {
+                XMethod xMethod = getXMethod();
+                try {
+                    if (isInherited(lock)) {
+                        inheritedLockObjects.put(lock, xMethod);
+                        if (isCollection(lock)) {
+                            inheritedCollectionLockObjects.add(lock, new BugInfo(this));
+                            try {
```

Comment:
I will restructure the `try` blocks to remove nested ones. I can simply handle both try blocks inside the `analyzeAssignmentsInHierarchy` with a minimal body. However, I have already restructured the other try block, so technically there are no nested `try` blocks, but I am still going to move them inside the function, because this way the driver code will be cleaner.